lib/schema/bolus.js

```diff
@@ -22,11 +22,66 @@ var util = require('util');
 var _ = require('lodash');
 
 var schema = require('./schema.js');
+var misc = require('../misc.js');
 
 var idFields = ['type', 'subType', 'deviceId', 'time'];
 schema.registerIdFields('bolus', idFields);
 
 module.exports = function(streamDAO){
+
+  function transformBoluses(data, cb) {
+    const previousIds = data.filter(datum => !!datum.previous).map(datum => {
+      const externalId = schema.generateId(datum.previous, idFields);
+      return streamDAO.generateInternalId(externalId, datum._groupId);
+    });
+    streamDAO.getDataIn(previousIds, (err, existingData) => {
+      if (err) {
+        return cb(err);
+      }
+
+      const {array, latestData, duplicates} = misc.removeDuplicates(existingData, data);
+      const finalData = {};
+
+      for (const datum of array) {
+        if (!datum.previous || ['injected', 'dual/square'].indexOf(datum.subType) > -1) {
+          const externalId = schema.generateId(datum, idFields);
+          const internalId = streamDAO.generateInternalId(externalId, datum._groupId);
+          finalData[internalId] = datum;
+          latestData[internalId] = datum;
+          continue;
+        }
+        const prevExternalId = schema.generateId(datum.previous, idFields);
+        const prevInternalId = streamDAO.generateInternalId(prevExternalId, datum._groupId);
+        let prev = latestData[prevInternalId];
+        // Todo logic depending on subType
```

Comment:
TODO?

Reply:
comment can be deleted, is implemented immediately below.